### PR TITLE
gh-68 - IterableFunction & Util to accept multiple Functions

### PR DESCRIPTION
**Pull request by m607123**
IterableFunction updated to accept multiple functions
IterableUtil updated to support changed functionality
Added and updated IterableFunction tests
Updated respective examples
Refactored examples to enable output of type iterable (resolves the contents rather than the object memory address)

### CODE REVIEW

**Comment by m607123** [comment]
Closing while changes made

File: core/src/main/java/uk/gov/gchq/koryphe/util/IterableUtil.java [hunk 1]
```diff
@@ -32,8 +34,12 @@ private IterableUtil() {
...
     }
 
-    public static <I_ITEM, O_ITEM> CloseableIterable<O_ITEM> map(final Iterable<I_ITEM> iterable, final Function<I_ITEM, O_ITEM> function) {
-        return new MappedIterable<>(iterable, function);
+    public static <I_ITEM, O_ITEM> CloseableIterable<O_ITEM> map(final Iterable<I_ITEM> iterable, final Function function) {
+        return map(iterable, Collections.singletonList(function));
+    }
+
+    public static <I_ITEM, O_ITEM> CloseableIterable<O_ITEM> map(final Iterable<I_ITEM> iterable, final List<Function> functions) {
+        return new MappedIterable<>(iterable, functions);
```
Comment (p013570) [inline core/src/main/java/uk/gov/gchq/koryphe/util/IterableUtil.java]: If functions is null or empty you could just return the functions.

Comment (p013570) [inline core/src/main/java/uk/gov/gchq/koryphe/util/IterableUtil.java]: If the iterable is null you should just return null.

File: core/src/main/java/uk/gov/gchq/koryphe/util/IterableUtil.java [hunk 2]
```diff
@@ -32,8 +34,12 @@ private IterableUtil() {
         // Empty
     }
 
-    public static <I_ITEM, O_ITEM> CloseableIterable<O_ITEM> map(final Iterable<I_ITEM> iterable, final Function<I_ITEM, O_ITEM> function) {
-        return new MappedIterable<>(iterable, function);
+    public static <I_ITEM, O_ITEM> CloseableIterable<O_ITEM> map(final Iterable<I_ITEM> iterable, final Function function) {
+        return map(iterable, Collections.singletonList(function));
```
Comment (p013570) [inline core/src/main/java/uk/gov/gchq/koryphe/util/IterableUtil.java]: I think we should check the function is not null.

File: core/src/main/java/uk/gov/gchq/koryphe/util/IterableUtil.java [hunk 3]
```diff
@@ -76,7 +82,18 @@ public boolean hasNext() {
 
         @Override
         public O_ITEM next() {
-            return function.apply(iterator.next());
+            Object item = iterator.next();
+            try {
+                for (final Function function : functions) {
+                    if (null == function) {
```
Comment (p013570) [inline core/src/main/java/uk/gov/gchq/koryphe/util/IterableUtil.java]: Perhaps move this (null == function) validation to the IterableUtil.map method.

File: core/src/test/java/uk/gov/gchq/koryphe/impl/function/IterableFunctionTest.java [hunk 4]
```diff
@@ -55,9 +57,9 @@ public void shouldJsonSerialiseAndDeserialise() throws IOException {
         // Then
         JsonSerialiser.assertEquals(String.format("{%n" +
                 "   \"class\" : \"uk.gov.gchq.koryphe.impl.function.IterableFunction\",%n" +
```
Comment (p013570) [inline core/src/test/java/uk/gov/gchq/koryphe/impl/function/IterableFunctionTest.java]: Perhaps update the test function to have multiple functions.

File: core/src/test/java/uk/gov/gchq/koryphe/impl/function/IterableFunctionTest.java [hunk 5]
```diff
@@ -92,4 +94,33 @@ public void shouldReturnEmptyIterableForEmptyInput() {
...
+    public void shouldPopulateFunctionFromBuilder() {
+        // Given
+        final IterableFunction<Integer, String> function = new IterableFunction.Builder<Integer>()
+                .first(Object::toString)
+                .then(Integer::valueOf)
+                .then(Object::toString)
+                .build();
+
+        // Then
+        assertEquals(3, function.getFunctions().size());
```
Comment (p013570) [inline core/src/test/java/uk/gov/gchq/koryphe/impl/function/IterableFunctionTest.java]: It would be good to test the actual functions values are as expected. For example, create variables for the functions (Object::toString, etc.) then:
```java
assertEquals(Arrays.asList(func1, func2, func3), function.getFunctions9));
```

File: core/src/test/java/uk/gov/gchq/koryphe/impl/function/IterableFunctionTest.java [hunk 6]
```diff
@@ -92,4 +94,33 @@ public void shouldReturnEmptyIterableForEmptyInput() {
...
+    }
+
+    @Test
+    public void shouldApplyMultipleFunctions() {
```
Comment (p013570) [inline core/src/test/java/uk/gov/gchq/koryphe/impl/function/IterableFunctionTest.java]: Please can you add some test for some of the other edge cases, like null functions, functions where the input/output types are incorrect, empty functions.
